cub/agent/agent_reduce.cuh

```diff
@@ -27,357 +27,418 @@
+
+  /// The input value type
+  using InputT = cub::detail::value_t<InputIteratorT>;
+
+  /// Vector type of InputT for data movement
+  using VectorT =
+    typename CubVector<InputT, AgentReducePolicy::VECTOR_LOAD_LENGTH>::Type;
+
+  /// Input iterator wrapper type (for applying cache modifier)
+  // Wrap the native input pointer with CacheModifiedInputIterator
+  // or directly use the supplied input iterator type
+  using WrappedInputIteratorT = cub::detail::conditional_t<
+    std::is_pointer<InputIteratorT>::value,
+    CacheModifiedInputIterator<AgentReducePolicy::LOAD_MODIFIER, InputT, OffsetT>,
+    InputIteratorT>;
+
+  /// Constants
+  static constexpr int BLOCK_THREADS      = AgentReducePolicy::BLOCK_THREADS;
+  static constexpr int ITEMS_PER_THREAD   = AgentReducePolicy::ITEMS_PER_THREAD;
+  static constexpr int TILE_ITEMS         = BLOCK_THREADS * ITEMS_PER_THREAD;
+  static constexpr int VECTOR_LOAD_LENGTH =
+    CUB_MIN(ITEMS_PER_THREAD, AgentReducePolicy::VECTOR_LOAD_LENGTH);
+
+  // Can vectorize according to the policy if the input iterator is a native
+  // pointer to a primitive type
+  static constexpr bool ATTEMPT_VECTORIZATION = (VECTOR_LOAD_LENGTH > 1) &&
+                            (ITEMS_PER_THREAD % VECTOR_LOAD_LENGTH == 0) &&
+                            (std::is_pointer<InputIteratorT>::value) &&
+                            Traits<InputT>::PRIMITIVE;
+
+  static constexpr CacheLoadModifier LOAD_MODIFIER =
+    AgentReducePolicy::LOAD_MODIFIER;
+
+  static constexpr BlockReduceAlgorithm BLOCK_ALGORITHM =
+    AgentReducePolicy::BLOCK_ALGORITHM;
+
+  /// Parameterized BlockReduce primitive
+  using BlockReduceT =
+    BlockReduce<AccumT, BLOCK_THREADS, AgentReducePolicy::BLOCK_ALGORITHM>;
+
+  /// Shared memory type required by this thread block
+  struct _TempStorage
+  {
+    typename BlockReduceT::TempStorage reduce;
+  };
+
+  /// Alias wrapper allowing storage to be unioned
+  struct TempStorage : Uninitialized<_TempStorage>
+  {};
+
+  //---------------------------------------------------------------------
+  // Per-thread fields
+  //---------------------------------------------------------------------
+
+  _TempStorage &temp_storage;         ///< Reference to temp_storage
+  InputIteratorT d_in;                ///< Input data to reduce
+  WrappedInputIteratorT d_wrapped_in; ///< Wrapped input data to reduce
+  ReductionOp reduction_op;           ///< Binary reduction operator
+
+  //---------------------------------------------------------------------
+  // Utility
+  //---------------------------------------------------------------------
+
+  // Whether or not the input is aligned with the vector type (specialized for
+  // types we can vectorize)
+  template <typename Iterator>
+  static __device__ __forceinline__ bool
+  IsAligned(Iterator d_in, Int2Type<true> /*can_vectorize*/)
+  {
+    return (size_t(d_in) & (sizeof(VectorT) - 1)) == 0;
```

Comment:
Shouldnt that be a property of the type `Iterator` and not the individual pointer `d_in`?

Reply:
I think the following code can illustrate what's going on here:
```cpp
#include <stdint.h>
#include <iostream>

template <class VectorT, class IteratorT>
bool is_aligned(IteratorT iter) {
  // can_vectorize = ... && is_pointer;
  return (size_t(iter) & (sizeof(VectorT) - 1)) == 0;
}

template <class VectorT, class IteratorT>
void foo(IteratorT iter) {
  std::cout << is_aligned<VectorT>(iter) << std::endl;
}

int main () {
  char data[32];
  foo<uint64_t>(data);
  foo<uint64_t>(data + 1);
}
```

How would you add property to the pointer? Are you suggesting wrapping pointers into a type that would perform `IsAligned` check in the constructor and provide a property to query?

Reply:
Urgh, yeah we are not comparing against the alignment of the pointer type but against 64bit alignment. Please disregard.